src/lib/comms/sol-lwm2m.c

```diff
@@ -2853,31 +2901,61 @@ static uint8_t
 handle_delete(struct sol_lwm2m_client *client,
     struct obj_ctx *obj_ctx, struct obj_instance *obj_instance)
 {
-    int r;
+    int r, ret = SOL_COAP_RESPONSE_CODE_BAD_REQUEST;
+    uint16_t i, j;
 
-    if (!obj_instance) {
-        SOL_WRN("Object instance was not provided to delete! (object id: %"
-            PRIu16 "", obj_ctx->obj->id);
-        return SOL_COAP_RESPONSE_CODE_BAD_REQUEST;
+    if (client->is_bootstrapping) {
+        clear_bootstrap_ctx(client);
     }
 
-    if (!obj_ctx->obj->del) {
-        SOL_WRN("The object %" PRIu16 " does not implement the delete method",
-            obj_ctx->obj->id);
-        return SOL_COAP_RESPONSE_CODE_NOT_ALLOWED;
-    }
+    //Specific Instance?
+    if (obj_ctx && obj_instance) {
+        if (!obj_ctx->obj->del) {
+            SOL_WRN("The object %" PRIu16 " does not implement the delete method",
+                obj_ctx->obj->id);
+            return client->is_bootstrapping ? SOL_COAP_RESPONSE_CODE_BAD_REQUEST :
+                   SOL_COAP_RESPONSE_CODE_NOT_ALLOWED;
+        }
 
-    r = obj_ctx->obj->del((void *)obj_instance->data,
-        (void *)client->user_data, client, obj_instance->id);
-    if (r < 0) {
-        SOL_WRN("Could not properly delete object id %"
-            PRIu16 " instance id: %" PRIu16 " reason:%d",
-            obj_ctx->obj->id, obj_instance->id, r);
-        return SOL_COAP_RESPONSE_CODE_NOT_ALLOWED;
+        r = obj_ctx->obj->del((void *)obj_instance->data,
+            (void *)client->user_data, client, obj_instance->id);
+        if (r < 0) {
+            SOL_WRN("Could not properly delete object id %"
+                PRIu16 " instance id: %" PRIu16 " reason:%d",
+                obj_ctx->obj->id, obj_instance->id, r);
+            return client->is_bootstrapping ? SOL_COAP_RESPONSE_CODE_BAD_REQUEST :
+                   SOL_COAP_RESPONSE_CODE_NOT_ALLOWED;
+        }
+
+        obj_instance->should_delete = true;
+        ret = SOL_COAP_RESPONSE_CODE_DELETED;
+    } else if (client->is_bootstrapping) {
+        SOL_VECTOR_FOREACH_IDX (&client->objects, obj_ctx, i) {
+            if (!obj_ctx->obj->del) {
+                SOL_WRN("The object %" PRIu16 " does not implement the delete method."
+                    " Skipping this Object.",
+                    obj_ctx->obj->id);
+                continue;
+            }
+
+            SOL_VECTOR_FOREACH_IDX (&obj_ctx->instances, obj_instance, j) {
+                r = obj_ctx->obj->del((void *)obj_instance->data,
+                    (void *)client->user_data, client, obj_instance->id);
+
+                if (r < 0) {
+                    SOL_WRN("Could not properly delete object id %"
+                        PRIu16 " instance id: %" PRIu16 " reason:%d",
+                        obj_ctx->obj->id, obj_instance->id, r);
```

Comment:
To the client?

I just did `ret = SOL_COAP_RESPONSE_CODE_BAD_REQUEST;`. It's a response to the Server.

Nice catch btw. :/